banjo/expression.cpp

```diff
@@ -37,4 +38,20 @@ make_requirements(Context& cxt,
   return cxt.make_requires(tparms, parms, reqs);
 }
 
+
+//TODO: Find a type for the Expr_list
+Expr&
+make_tuple_expr(Context& cxt, Expr_list& l)
+{
+  Type_list tlist;
+  for(auto eli = l.begin(); eli != l.end(); eli++) {
+    tlist.push_back(eli->type());
+  }
+  
+  Tuple_type t(tlist);
+  
+  return cxt.make_tuple_expr(t, l);
```

Comment:
I think I fixed it. Let me know if that's not right.
On May 4, 2016 10:34 AM, "Andrew Sutton" <notifications@github.com> wrote:

> In banjo/expression.cpp
> <https://github.com/asutton/banjo/pull/14#discussion_r62049332>:
>
> > @@ -37,4 +38,20 @@ make_requirements(Context& cxt,
> >    return cxt.make_requires(tparms, parms, reqs);
> >  }
> >
> > +
> > +//TODO: Find a type for the Expr_list
> > +Expr&
> > +make_tuple_expr(Context& cxt, Expr_list& l)
> > +{
> > +  Type_list tlist;
> > +  for(auto eli = l.begin(); eli != l.end(); eli++) {
> > +    tlist.push_back(eli->type());
> > +  }
> > +
> > +  Tuple_type t(tlist);
> > +
> > +  return cxt.make_tuple_expr(t, l);
>
> t is a local variable. You are passing as a reference to make_tuple_expr,
> which dynamically allocates a new object having a pointer to t. You return
> that object. t is destroyed. The pointer is invalid.
> On Wed, May 4, 2016 at 9:59 AM cjb129 ***@***.***> wrote: In
> banjo/expression.cpp <#14 (comment)
> <https://github.com/asutton/banjo/pull/14#discussion_r62043086>>: > @@
> -37,4 +38,20 @@ make_requirements(Context& cxt, > return
> cxt.make_requires(tparms, parms, reqs); > } > > + > +//TODO: Find a type
> for the Expr_list > +Expr& > +make_tuple_expr(Context& cxt, Expr_list& l) >
> +{ > + Type_list tlist; > + for(auto eli = l.begin(); eli != l.end();
> eli++) { > + tlist.push_back(eli->type()); > + } > + > + Tuple_type
> t(tlist); > + > + return cxt.make_tuple_expr(t, l); How do I test for that
> segfault? Are you referring to tlist, or t, or both? Are you saying I
> shouldn't be using any local variables? Thanks … <#m_-8656248427600988747_>
> On Wed, May 4, 2016 at 9:15 AM, Andrew Sutton ***@***.***> wrote: In
> banjo/expression.cpp <#14 <https://github.com/asutton/banjo/pull/14>
> (comment) <#14 (comment)
> <https://github.com/asutton/banjo/pull/14#discussion_r62036058>>>: > @@
> -37,4 +38,20 @@ make_requirements(Context& cxt, > return
> cxt.make_requires(tparms, parms, reqs); > } > > + > +//TODO: Find a type
> for the Expr_list > +Expr& > +make_tuple_expr(Context& cxt, Expr_list& l) >
> +{ > + Type_list tlist; > + for(auto eli = l.begin(); eli != l.end();
> eli++) { > + tlist.push_back(eli->type()); > + } > + > + Tuple_type
> t(tlist); > + > + return cxt.make_tuple_expr(t, l); This will segfault. You
> are leaking a reference to a local variable. — You are receiving this
> because you authored the thread. Reply to this email directly or view it on
> GitHub <
> https://github.com/asutton/banjo/pull/14/files/abf73a25543526046288837c2f637be7edb9c535#r62036058>
> — You are receiving this because you commented. Reply to this email
> directly or view it on GitHub <
> https://github.com/asutton/banjo/pull/14/files/abf73a25543526046288837c2f637be7edb9c535#r62043086>
>
> -- Andrew Sutton
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/asutton/banjo/pull/14/files/abf73a25543526046288837c2f637be7edb9c535#r62049332>
>